In [1]:
import pandas as pd
import numpy as np
import glob

In [2]:
eQTL = pd.read_csv('/data/users2/pnadigapusuresh1/Projects/ukbiobank/Data/QTL/eQTL_hg19_1FPKM_with_rs_id.csv')

In [3]:
eQTL

Unnamed: 0          gene_id gene_chr  gene_start  gene_end strand  \
0                 0  ENSG00000204316    chr17    73901494  73901494      +   
1                 1  ENSG00000204316    chr17    73901494  73901494      +   
2                 2  ENSG00000204316    chr17    73901494  73901494      +   
3                 3  ENSG00000204316    chr17    73901494  73901494      +   
4                 4  ENSG00000204316    chr17    73901494  73901494      +   
...             ...              ...      ...         ...       ...    ...   
2120746     2120746  ENSG00000186532    chr17     1733175   1733175      +   
2120747     2120747  ENSG00000186532    chr17     1733175   1733175      +   
2120748     2120748  ENSG00000186532    chr17     1733175   1733175      +   
2120749     2120749  ENSG00000186532    chr17     1733175   1733175      +   
2120750     2120750  ENSG00000185924    chr17     1928639   1928639      +   

         number_of_SNPs_tested  SNP_distance_to_TSS       SNP_id SNP_chr  \
0                         3394              -955836  17:72945658   chr17   
1                         3394              -546694  17:73354800   chr17   
2                         3394              -118750  17:73782744   chr17   
3                         3394               -59792  17:73841702   chr17   
4                         3394               -54978  17:73846516   chr17   
...                        ...                  ...          ...     ...   
2120746                   4191                48815   17:1781990   chr17   
2120747                   4191                49967   17:1783142   chr17   
2120748                   4191                51643   17:1784818   chr17   
2120749                   4191                53551   17:1786726   chr17   
2120750                   4336               -16641   17:1911998   chr17   

         SNP_start   SNP_end  nominal_pval  regression_slope  top_SNP  \
0         72945658  72945658      0.000852          0.057756        0   
1         73354800  73354800      0.001217         -0.099736        0   
2         73782744  73782744      0.000407         -0.097620        0   
3         73841702  73841702      0.001396         -0.041055        0   
4         73846516  73846516      0.000189         -0.053441        0   
...            ...       ...           ...               ...      ...   
2120746    1781990   1781990      0.000035          0.109385        0   
2120747    1783142   1783142      0.000035          0.109502        0   
2120748    1784818   1784818      0.000033          0.110230        0   
2120749    1786726   1786726      0.000066          0.106669        0   
2120750    1911998   1911998      0.001325          0.032620        1   

              FDR      snps         rs_id  
0        0.029260  72945658    rs77018335  
1        0.039173  73354800  rs1237746321  
2        0.015733  73782744     rs8077281  
3        0.043746  73841702     rs9900122  
4        0.008126  73846516     rs2004452  
...           ...       ...           ...  
2120746  0.001832   1781990     rs4638622  
2120747  0.001831   1783142    rs12948780  
2120748  0.001736   1784818    rs12943171  
2120749  0.003242   1786726    rs11872052  
2120750  0.041942   1911998    rs72809519  

[2120751 rows x 18 columns]

In [7]:
eQTL.head()

Unnamed: 0          gene_id gene_chr  gene_start  gene_end strand  \
0           0  ENSG00000204316    chr17    73901494  73901494      +   
1           1  ENSG00000204316    chr17    73901494  73901494      +   
2           2  ENSG00000204316    chr17    73901494  73901494      +   
3           3  ENSG00000204316    chr17    73901494  73901494      +   
4           4  ENSG00000204316    chr17    73901494  73901494      +   

   number_of_SNPs_tested  SNP_distance_to_TSS       SNP_id SNP_chr  SNP_start  \
0                   3394              -955836  17:72945658   chr17   72945658   
1                   3394              -546694  17:73354800   chr17   73354800   
2                   3394              -118750  17:73782744   chr17   73782744   
3                   3394               -59792  17:73841702   chr17   73841702   
4                   3394               -54978  17:73846516   chr17   73846516   

    SNP_end  nominal_pval  regression_slope  top_SNP       FDR      snps  \
0  72945658      0.000852          0.057756        0  0.029260  72945658   
1  73354800      0.001217         -0.099736        0  0.039173  73354800   
2  73782744      0.000407         -0.097620        0  0.015733  73782744   
3  73841702      0.001396         -0.041055        0  0.043746  73841702   
4  73846516      0.000189         -0.053441        0  0.008126  73846516   

          rs_id  
0    rs77018335  
1  rs1237746321  
2     rs8077281  
3     rs9900122  
4     rs2004452

#### snps selected after LD pruning using 0.1 treshold

In [8]:
pruned_snps = glob.glob('*.in')

In [9]:
snps_pruned_in = [pd.read_csv(file,header=None) for file in pruned_snps]

In [10]:
all_snps_pruned_in = pd.concat(snps_pruned_in,axis=0)

In [11]:
len(eQTL[eQTL['rs_id'].isin(all_snps_pruned_in[0])]['gene_id'].unique())

12886

In [12]:
eQTL.shape

(2120751, 18)

In [13]:
np.savetxt('snp_id.txt',eQTL.SNP_id.values,fmt='%s')

In [14]:
snp_ids = eQTL.SNP_id.to_list()

In [29]:
res = map(lambda x: x.split(':')[1],snp_ids)

In [30]:
only_snp_ids = list(res)

In [31]:
eQTL['snps'] = only_snp_ids

In [11]:
ukb_snps = pd.read_csv(filepath_or_buffer='/data/users2/pnadigapusuresh1/Projects/ukbiobank/Data/ukb_cal_imggen_3_recode.raw',nrows=1,delimiter=' ')

In [12]:
len(ukb_snps.columns)

554865

In [13]:
ukb_snps.loc[:,'FID':'rs3131962_A']

FID      IID  PAT  MAT  SEX  PHENOTYPE  rs3131962_A
0  1000293  1000293    0    0    2         -9            0

In [22]:
snp_ids_ukb = ukb_snps.columns.tolist()

In [23]:
snp_ids_ukb[1:100]

['IID',
 'PAT',
 'MAT',
 'SEX',
 'PHENOTYPE',
 'rs3131962_A',
 'rs12562034_A',
 'rs4040617_G',
 'rs79373928_G',
 'rs11240779_G',
 'rs57181708_G',
 'rs4422948_G',
 'rs4970383_A',
 'rs4970382_C',
 'rs950122_C',
 'rs6657440_C',
 'rs72631889_T',
 'rs4970459_T',
 'rs13303101_A',
 'rs192998324_G',
 'rs55678698_T',
 'rs74047407_A',
 'rs77579483_T',
 'rs1110052_G',
 'Affx-52325915_GA',
 'rs3748592_A',
 'rs3748593_A',
 'rs2272757_G',
 'rs35471880_A',
 'rs67274836_A',
 'rs114631528_A',
 'rs3748597_T',
 'rs13303065_C',
 'rs112164716_C',
 'rs13303010_G',
 'rs6696971_T',
 'rs6696609_T',
 'rs41285810_T',
 'rs3829740_G',
 'Affx-15504485_A',
 'rs3748588_T',
 'rs2340596_G',
 'rs6659460_T',
 'rs28469033_C',
 'rs28869591_T',
 'rs6677131_T',
 'rs78164078_A',
 'rs3128109_T',
 'rs35940137_A',
 'rs76233940_A',
 'Affx-15871198_A',
 'rs2799070_C',
 'rs2710890_G',
 'rs28591569_G',
 'rs28670633_T',
 'rs13303147_T',
 'rs4970349_C',
 'rs2465130_C',
 'rs138288952_A',
 'Affx-52128261_A',
 'rs2275811_C',
 'rs2710887_

In [38]:
cd '/data/analysis/collaboration/Multi-sites_genetics/GeneticDB/GenomeBuild/human_9606_b151_GRCh37p13'

/data/analysis/collaboration/Multi-sites_genetics/GeneticDB/GenomeBuild/human_9606_b151_GRCh37p13


In [39]:
df_all = pd.read_csv('00-All_chrpos_rsid.txt',delimiter='\t\t\t',header=None)

/tmp/ipykernel_120931/3369962242.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  df_all = pd.read_csv('00-All_chrpos_rsid.txt',delimiter='\t\t\t',header=None)


In [40]:
df_all.head()

0             1
0  1:10019   rs775809821
1  1:10039   rs978760828
2  1:10043  rs1008829651
3  1:10051  rs1052373574
4  1:10051  rs1326880612

In [41]:
snps = eQTL.SNP_id

In [42]:
df_all.columns = ['chr','rs']

In [43]:
df_all.head()

chr            rs
0  1:10019   rs775809821
1  1:10039   rs978760828
2  1:10043  rs1008829651
3  1:10051  rs1052373574
4  1:10051  rs1326880612

In [44]:
df_all.set_index('chr')

rs
chr                   
1:10019    rs775809821
1:10039    rs978760828
1:10043   rs1008829651
1:10051   rs1052373574
1:10051   rs1326880612
...                ...
26:16512   rs373943637
26:16519     rs3937033
26:16526   rs386829315
26:16527   rs370705831
26:16529   rs879492040

[638388005 rows x 1 columns]

In [45]:
df_all['chr'].head()

0    1:10019
1    1:10039
2    1:10043
3    1:10051
4    1:10051
Name: chr, dtype: object

In [46]:
df_all = df_all.iloc[:615337877]

In [47]:
pd.Series(df_all.rs.values,index=df_all.chr).to_dict()

{'1:10019': 'rs775809821',
 '1:10039': 'rs978760828',
 '1:10043': 'rs1008829651',
 '1:10051': 'rs1326880612',
 '1:10055': 'rs892501864',
 '1:10063': 'rs1010989343',
 '1:10067': 'rs1489251879',
 '1:10077': 'rs1022805358',
 '1:10108': 'rs1377973775',
 '1:10109': 'rs1462685959',
 '1:10119': 'rs1156821933',
 '1:10120': 'rs1390810297',
 '1:10122': 'rs1199085795',
 '1:10128': 'rs1457723673',
 '1:10131': 'rs1289482855',
 '1:10132': 'rs1390118706',
 '1:10134': 'rs1385251551',
 '1:10135': 'rs1303755152',
 '1:10137': 'rs1346539515',
 '1:10138': 'rs1228214171',
 '1:10139': 'rs368469931',
 '1:10144': 'rs144773400',
 '1:10146': 'rs779258992',
 '1:10147': 'rs1292226269',
 '1:10149': 'rs1286868604',
 '1:10150': 'rs371194064',
 '1:10153': 'rs1331781659',
 '1:10154': 'rs1240389757',
 '1:10155': 'rs1264289758',
 '1:10156': 'rs1487252449',
 '1:10159': 'rs1211258708',
 '1:10160': 'rs1269202771',
 '1:10162': 'rs1482594023',
 '1:10163': 'rs1179689498',
 '1:10164': 'rs1413947121',
 '1:10165': 'rs796884232',


In [49]:
d = _

In [50]:
eQTL.SNP_id.map(d)

0            rs77018335
1          rs1237746321
2             rs8077281
3             rs9900122
4             rs2004452
               ...     
2120746       rs4638622
2120747      rs12948780
2120748      rs12943171
2120749      rs11872052
2120750      rs72809519
Name: SNP_id, Length: 2120751, dtype: object

In [51]:
snps = _

In [53]:
eQTL['rs_id'] = snps.values

### Saving the eQTL with their rs_id

In [ ]:
eQTL.to_csv('./eQTL_hg19_1FPKM_with_rs_id.csv')

In [18]:
snps.to_csv('rs_snps_eqtl.txt',index=True)

In [36]:
snps = pd.read_csv('./rs_snps_eqtl.txt')

In [8]:
final_snps = pd.read_csv('../ukb_cal_imggen_eQTL_matched_recode.raw',delim_whitespace=True,nrows=3)

In [9]:
final_snps.shape

(3, 28654)

In [10]:
final_snps.columns

Index(['FID', 'IID', 'PAT', 'MAT', 'SEX', 'PHENOTYPE', 'rs3128109_T',
       'rs4970349_C', 'rs2799072_C', 'rs11260596_T',
       ...
       'rs6151416_A', 'rs5770964_G', 'rs117063829_T', 'rs6010045_T',
       'rs36006314_G', 'rs2157711_C', 'rs5770821_G', 'rs78342408_T',
       'rs56207913_T', 'rs73174437_T'],
      dtype='object', length=28654)

In [2]:
for i in range(1,23):
    print(f'ukb_imputed_imggen_chr{i}')

ukb_imputed_imggen_chr1
ukb_imputed_imggen_chr2
ukb_imputed_imggen_chr3
ukb_imputed_imggen_chr4
ukb_imputed_imggen_chr5
ukb_imputed_imggen_chr6
ukb_imputed_imggen_chr7
ukb_imputed_imggen_chr8
ukb_imputed_imggen_chr9
ukb_imputed_imggen_chr10
ukb_imputed_imggen_chr11
ukb_imputed_imggen_chr12
ukb_imputed_imggen_chr13
ukb_imputed_imggen_chr14
ukb_imputed_imggen_chr15
ukb_imputed_imggen_chr16
ukb_imputed_imggen_chr17
ukb_imputed_imggen_chr18
ukb_imputed_imggen_chr19
ukb_imputed_imggen_chr20
ukb_imputed_imggen_chr21
ukb_imputed_imggen_chr22


## Creating the ADJ matrices

In [4]:
import glob

In [5]:
csv_flies = glob.glob('/data/users2/pnadigapusuresh1/Projects/ukbiobank/Data/QTL'+'/*pruned_recoded.raw')

In [6]:
df_list = [pd.read_csv(file,sep=' ',nrows=3,index_col='FID',usecols=lambda x: x not in ['PHENOTYPE','SEX','IID','PAT','MAT']) for file in csv_flies]

In [7]:
df_list[20].iloc[:,:10]

rs13437666_C  rs79146206_G  rs6964562_C  rs4593475_C  rs6968087_A  \
FID                                                                          
1767364             0             0            2            2            0   
2268235             0             0            2            2            1   
4521147             0             0            0            0            0   

         rs28378990_A  rs6945181_A  rs148955469_G  rs117084377_C  rs75418879_T  
FID                                                                             
1767364             0            1              0              0             0  
2268235             1            0              0              0             0  
4521147             0            1              0              0             0

In [8]:
big_df = pd.concat(df_list,axis=1)

In [9]:
big_df.shape

(3, 37606)

In [10]:
big_df.iloc[:,:10]

rs77114150_A  rs36020100_A  rs28648766_T  rs441438_C  rs331007_G  \
FID                                                                         
1767364             0             2             0           1           1   
2268235             0             1             0           1           2   
4521147             0             0             0           0           1   

         rs61614253_A  rs7225407_A  rs115874046_A  rs1570000_C  rs78300101_C  
FID                                                                           
1767364             0            1              0            0             0  
2268235             0            0              1            0             0  
4521147             0            0              0            0             0

In [11]:
final_snps = big_df.columns.tolist()

In [12]:
final_snps = list(map(lambda snp: snp.split('_')[0],final_snps))

In [13]:
final_eqtls = eQTL[eQTL['rs_id'].isin(final_snps)]

In [14]:
final_eqtls

Unnamed: 0          gene_id gene_chr  gene_start  gene_end strand  \
33               33  ENSG00000204316    chr17    73901494  73901494      +   
41               41  ENSG00000188878    chr17    73937119  73937119      +   
120             120  ENSG00000257949    chr17    73975301  73975301      +   
136             136  ENSG00000257949    chr17    73975301  73975301      +   
137             137  ENSG00000257949    chr17    73975301  73975301      +   
...             ...              ...      ...         ...       ...    ...   
2120601     2120601  ENSG00000132386    chr17     1665253   1665253      +   
2120610     2120610  ENSG00000132383    chr17     1732996   1732996      +   
2120696     2120696  ENSG00000132383    chr17     1732996   1732996      +   
2120702     2120702  ENSG00000186532    chr17     1733175   1733175      +   
2120750     2120750  ENSG00000185924    chr17     1928639   1928639      +   

         number_of_SNPs_tested  SNP_distance_to_TSS       SNP_id SNP_chr  \
33                        3394                40702  17:73942196   chr17   
41                        3477              -776304  17:73160815   chr17   
120                       3544                11913  17:73987214   chr17   
136                       3544                34766  17:74010067   chr17   
137                       3544                34971  17:74010272   chr17   
...                        ...                  ...          ...     ...   
2120601                   4233               658410   17:2323663   chr17   
2120610                   4194              -429894   17:1303102   chr17   
2120696                   4194                67221   17:1800217   chr17   
2120702                   4191              -774661    17:958514   chr17   
2120750                   4336               -16641   17:1911998   chr17   

         SNP_start   SNP_end  nominal_pval  regression_slope  top_SNP  \
33        73942196  73942196  3.908310e-05          0.131688        0   
41        73160815  73160815  4.038930e-05          0.346087        0   
120       73987214  73987214  1.004390e-09         -0.352123        0   
136       74010067  74010067  5.757640e-06         -0.429004        0   
137       74010272  74010272  4.869350e-11          0.126632        0   
...            ...       ...           ...               ...      ...   
2120601    2323663   2323663  7.486830e-04         -0.308176        0   
2120610    1303102   1303102  6.912490e-04         -0.134710        0   
2120696    1800217   1800217  3.102950e-09         -0.104906        1   
2120702     958514    958514  5.168770e-04         -0.115433        0   
2120750    1911998   1911998  1.324730e-03          0.032620        1   

                  FDR      snps        rs_id  
33       2.023840e-03  73942196   rs71382188  
41       2.084354e-03  73160815  rs138046159  
120      1.132131e-07  73987214  rs139824587  
136      3.586709e-04  74010067  rs112572575  
137      6.435118e-09  74010272    rs2598420  
...               ...       ...          ...  
2120601  2.627355e-02   2323663   rs61753102  
2120610  2.458446e-02   1303102   rs35172388  
2120696  3.282862e-07   1800217    rs7406062  
2120702  1.925770e-02    958514    rs2261299  
2120750  4.194184e-02   1911998   rs72809519  

[45455 rows x 18 columns]

In [15]:
len(final_eqtls['gene_id'].unique())

12886

In [16]:
len(final_eqtls['SNP_id'].unique())

37470

In [17]:
final_eqtls.groupby(['SNP_id','gene_id']).size()

SNP_id        gene_id        
10:100020572  ENSG00000166024    1
10:100041538  ENSG00000119943    1
              ENSG00000155265    1
10:100045000  ENSG00000171160    1
10:100075076  ENSG00000119943    1
                                ..
9:99983931    ENSG00000188801    1
              ENSG00000203279    1
              ENSG00000213612    1
              ENSG00000228038    1
              ENSG00000255036    1
Length: 45455, dtype: int64

In [18]:
final_eqtls[final_eqtls.duplicated(['SNP_id'])]

Unnamed: 0          gene_id gene_chr  gene_start  gene_end strand  \
293             293  ENSG00000167881    chr17    74068734  74068734      +   
333             333  ENSG00000182473    chr17    74079095  74079095      +   
525             525  ENSG00000129646    chr17    74303761  74303761      +   
567             567  ENSG00000129646    chr17    74303761  74303761      +   
581             581  ENSG00000129646    chr17    74303761  74303761      +   
...             ...              ...      ...         ...       ...    ...   
2120417     2120417  ENSG00000167703    chr17     1532130   1532130      +   
2120441     2120441  ENSG00000074660    chr17     1549041   1549041      +   
2120449     2120449  ENSG00000167705    chr17     1552326   1552326      +   
2120586     2120586  ENSG00000132386    chr17     1665253   1665253      +   
2120595     2120595  ENSG00000132386    chr17     1665253   1665253      +   

         number_of_SNPs_tested  SNP_distance_to_TSS       SNP_id SNP_chr  \
293                       3651                22456  17:74091190   chr17   
333                       3682                12095  17:74091190   chr17   
525                       3881              -316547  17:73987214   chr17   
567                       3881               -53755  17:74250006   chr17   
581                       3881               -32903  17:74270858   chr17   
...                        ...                  ...          ...     ...   
2120417                   4345               925514   17:2457644   chr17   
2120441                   4332               729603   17:2278644   chr17   
2120449                   4334              -195682   17:1356644   chr17   
2120586                   4233                66344   17:1731597   chr17   
2120595                   4233               135696   17:1800949   chr17   

         SNP_start   SNP_end  nominal_pval  regression_slope  top_SNP  \
293       74091190  74091190  1.695810e-07          0.051636        0   
333       74091190  74091190  2.126470e-05          0.065754        0   
525       73987214  73987214  3.664650e-08         -0.307343        0   
567       74250006  74250006  1.227960e-15         -0.143533        0   
581       74270858  74270858  2.711150e-07         -0.201246        0   
...            ...       ...           ...               ...      ...   
2120417    2457644   2457644  9.269000e-04          0.202301        0   
2120441    2278644   2278644  7.940860e-07         -0.103760        1   
2120449    1356644   1356644  1.609350e-03          0.144334        0   
2120586    1731597   1731597  1.507050e-04          0.349130        0   
2120595    1800949   1800949  2.987730e-09          0.150302        0   

                  FDR      snps        rs_id  
293      1.393950e-05  74091190    rs7212071  
333      1.171846e-03  74091190    rs7212071  
525      3.343890e-06  73987214  rs139824587  
567      2.521036e-13  74250006    rs8071545  
581      2.156326e-05  74270858   rs74808805  
...               ...       ...          ...  
2120417  3.135794e-02   2457644  rs117297421  
2120441  5.829612e-05   2278644    rs2252910  
2120449  4.902310e-02   1356644  rs117985867  
2120586  6.673966e-03   1731597  rs140729163  
2120595  3.167279e-07   1800949    rs3744766  

[7985 rows x 18 columns]

#### Loading Alzhiemers protien ID's

In [19]:
p_g = pd.read_csv('/data/users2/pnadigapusuresh1/Projects/ukbiobank/bio_informed_nn/gProfiler_hsapiens_6-20-2023_4-00-56 PM.csv')

In [20]:
p_g.columns

Index(['initial_alias', 'converted_alias', 'name', 'description', 'namespace'], dtype='object')

In [21]:
p_g_new = p_g.drop_duplicates(subset=['initial_alias','converted_alias'])

In [22]:
p_g_new

initial_alias  converted_alias    name  \
0           P05067  ENSG00000142192     APP   
30          P49768  ENSG00000080815   PSEN1   
104         P54253  ENSG00000124788   ATXN1   
151         P42858  ENSG00000197386     HTT   
270         O14672  ENSG00000137845  ADAM10   
...            ...              ...     ...   
3237        P43116  ENSG00000125384  PTGER2   
3245        P53805  ENSG00000159200   RCAN1   
3264        Q96L34  ENSG00000007047   MARK4   
3269        Q7KZI7  ENSG00000072518   MARK2   
3271        Q9P0L2  ENSG00000116141   MARK1   

                                            description  \
0     amyloid beta precursor protein [Source:HGNC Sy...   
30      presenilin 1 [Source:HGNC Symbol;Acc:HGNC:9508]   
104        ataxin 1 [Source:HGNC Symbol;Acc:HGNC:10548]   
151       huntingtin [Source:HGNC Symbol;Acc:HGNC:4851]   
270   ADAM metallopeptidase domain 10 [Source:HGNC S...   
...                                                 ...   
3237  prostaglandin E receptor 2 [Source:HGNC Symbol...   
3245  regulator of calcineurin 1 [Source:HGNC Symbol...   
3264  microtubule affinity regulating kinase 4 [Sour...   
3269  microtubule affinity regulating kinase 2 [Sour...   
3271  microtubule affinity regulating kinase 1 [Sour...   

                                namespace  
0     UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
30    UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
104   UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
151   UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
270   UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
...                                   ...  
3237  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
3245  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
3264  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
3269  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
3271  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  

[225 rows x 5 columns]

In [23]:
len(p_g_new['converted_alias'].unique())

214

In [24]:
alzhimers_gene_id = p_g_new['converted_alias'].unique().tolist()

#### Matching final_eqtl gene_id with the alzhiemer's gene id

In [161]:
first_second_level_genes = final_eqtls[final_eqtls['gene_id'].isin(alzhimers_gene_id)]

In [162]:
first_second_level_genes

Unnamed: 0          gene_id gene_chr  gene_start   gene_end strand  \
19254         19254  ENSG00000175582    chr11    73472182   73472182      +   
19308         19308  ENSG00000175582    chr11    73472182   73472182      +   
21876         21876  ENSG00000137486    chr11    75062749   75062749      +   
21957         21957  ENSG00000149243    chr11    75141236   75141236      +   
22000         22000  ENSG00000149243    chr11    75141236   75141236      +   
...             ...              ...      ...         ...        ...    ...   
2111780     2111780  ENSG00000175166     chr3   184016497  184016497      +   
2111782     2111782  ENSG00000175166     chr3   184016497  184016497      +   
2111785     2111785  ENSG00000175166     chr3   184016497  184016497      +   
2111788     2111788  ENSG00000175166     chr3   184016497  184016497      +   
2111791     2111791  ENSG00000175166     chr3   184016497  184016497      +   

         number_of_SNPs_tested  SNP_distance_to_TSS       SNP_id SNP_chr  \
19254                     3320               115671  11:73587853   chr11   
19308                     3320               233082  11:73705264   chr11   
21876                     3105              -654643  11:74408106   chr11   
21957                     3064               -95693  11:75045543   chr11   
22000                     3064               -14506  11:75126730   chr11   
...                        ...                  ...          ...     ...   
2111780                   4337              -174724  3:183841773    chr3   
2111782                   4337               -77640  3:183938857    chr3   
2111785                   4337                11042  3:184027539    chr3   
2111788                   4337                69239  3:184085736    chr3   
2111791                   4337               350532  3:184367029    chr3   

         SNP_start    SNP_end  nominal_pval  regression_slope  top_SNP  \
19254     73587853   73587853      0.000721         -0.168627        0   
19308     73705264   73705264      0.000071          0.061784        0   
21876     74408106   74408106      0.001491         -0.091301        0   
21957     75045543   75045543      0.000402          0.225383        0   
22000     75126730   75126730      0.000296          0.054877        0   
...            ...        ...           ...               ...      ...   
2111780  183841773  183841773      0.000086          0.167709        0   
2111782  183938857  183938857      0.000948         -0.038527        0   
2111785  184027539  184027539      0.000002         -0.056150        1   
2111788  184085736  184085736      0.001154         -0.071793        0   
2111791  184367029  184367029      0.000443          0.102622        0   

              FDR       snps        rs_id  
19254    0.025454   73587853   rs76332820  
19308    0.003427   73705264    rs4944873  
21876    0.046117   74408106    rs7131683  
21957    0.015567   75045543  rs139848901  
22000    0.011975   75126730   rs10793118  
...           ...        ...          ...  
2111780  0.004072  183841773   rs77980659  
2111782  0.031945  183938857    rs4447797  
2111785  0.000106  184027539   rs13062950  
2111788  0.037505  184085736    rs9839511  
2111791  0.016906  184367029  rs145914525  

[401 rows x 18 columns]

In [163]:
second_level_matched_genes = first_second_level_genes['gene_id'].unique().tolist()

In [164]:
len(second_level_matched_genes)

133

##### First level SNPs and 2nd level Genes

In [165]:
len(first_second_level_genes['SNP_id'].tolist())

401

In [166]:
first_second_level_genes[first_second_level_genes['SNP_id'].duplicated()]

Unnamed: 0          gene_id gene_chr  gene_start  gene_end strand  \
1537290     1537290  ENSG00000177697    chr11      832843    832843      +   

         number_of_SNPs_tested  SNP_distance_to_TSS      SNP_id SNP_chr  \
1537290                   3978               209124  11:1041967   chr11   

         SNP_start  SNP_end  nominal_pval  regression_slope  top_SNP  \
1537290    1041967  1041967      0.000017         -0.071387        0   

              FDR     snps       rs_id  
1537290  0.000972  1041967  rs72842424

In [167]:
first_second_level_genes[first_second_level_genes['SNP_id'] == '11:1041967']

Unnamed: 0          gene_id gene_chr  gene_start  gene_end strand  \
1466290     1466290  ENSG00000023191    chr11      503570    503570      +   
1537290     1537290  ENSG00000177697    chr11      832843    832843      +   

         number_of_SNPs_tested  SNP_distance_to_TSS      SNP_id SNP_chr  \
1466290                   3113               538397  11:1041967   chr11   
1537290                   3978               209124  11:1041967   chr11   

         SNP_start  SNP_end  nominal_pval  regression_slope  top_SNP  \
1466290    1041967  1041967      0.000501          0.049051        0   
1537290    1041967  1041967      0.000017         -0.071387        0   

              FDR     snps       rs_id  
1466290  0.018756  1041967  rs72842424  
1537290  0.000972  1041967  rs72842424

In [172]:
adj_matrix_first_second = pd.crosstab(first_second_level_genes['SNP_id'],first_second_level_genes['gene_id'])

In [173]:
adj_matrix_first_second.shape

(400, 133)

In [57]:
with open('first_second_adj_matrix.npy','wb') as f:
    np.save(f, adj_matrix_first_second.to_numpy())


#### Mstching these matched genes between eqtl's and alzhiemers genes for third level connections

In [174]:
gene_dataset = pd.read_csv('2023-06-16-10-44.txt',delimiter='\t')

In [175]:
gene_dataset

# ID(s) interactor A ID(s) interactor B  \
0        uniprotkb:P05067   uniprotkb:O95864   
1        uniprotkb:P05067   uniprotkb:P13861   
2        uniprotkb:P05067   uniprotkb:Q8TD31   
3        uniprotkb:P05067   uniprotkb:Q15392   
4        uniprotkb:P05067   uniprotkb:Q15125   
...                   ...                ...   
2395     uniprotkb:P05067   uniprotkb:P63104   
2396     uniprotkb:P05067   uniprotkb:P07900   
2397     uniprotkb:P05067   uniprotkb:P46459   
2398     uniprotkb:P10636   uniprotkb:P27448   
2399     uniprotkb:P05067   uniprotkb:Q05193   

                                Alt. ID(s) interactor A  \
0     ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
1     ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
2     ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
3     ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
4     ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
...                                                 ...   
2395  ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
2396  ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
2397  ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
2398  uniprotkb:P18518|uniprotkb:Q14799|uniprotkb:Q1...   
2399  ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   

                                Alt. ID(s) interactor B  \
0     uniprotkb:A8K2M6|uniprotkb:Q6MZQ7|uniprotkb:Q9...   
1     uniprotkb:Q16823|intact:EBI-6116682|uniprotkb:...   
2     uniprotkb:Q9NRK8|uniprotkb:Q9NWY9|uniprotkb:Q9...   
3     uniprotkb:D3DQ51|uniprotkb:Q9HBA8|uniprotkb:B7...   
4     uniprotkb:Q6IBI9|uniprotkb:Q6FGL3|ensembl:ENSP...   
...                                                 ...   
2395  uniprotkb:P29213|uniprotkb:P29312|uniprotkb:Q3...   
2396  intact:EBI-1642144|uniprotkb:Q2PP14|ensembl:EN...   
2397  uniprotkb:Q8N6D7|uniprotkb:Q9UKZ2|uniprotkb:A8...   
2398  uniprotkb:O60219|uniprotkb:Q8TB41|uniprotkb:Q8...   
2399  uniprotkb:A6NLM6|uniprotkb:Q5SYX0|uniprotkb:Q5...   

                                 Alias(es) interactor A  \
0     psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
1     psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
2     psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
3     psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
4     psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
...                                                 ...   
2395  psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
2396  psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
2397  psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
2398  psi-mi:tau_human(display_short)|psi-mi:Microtu...   
2399  psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   

                                 Alias(es) interactor B  \
0     psi-mi:fads2_human(display_short)|psi-mi:Acyl-...   
1     psi-mi:kap2_human(display_short)|psi-mi:cAMP-d...   
2     psi-mi:cchcr_human(display_short)|psi-mi:Coile...   
3     psi-mi:dhc24_human(display_short)|psi-mi:"Delt...   
4     psi-mi:ebp_human(display_short)|psi-mi:"3-beta...   
...                                                 ...   
2395  psi-mi:1433z_human(display_short)|psi-mi:14-3-...   
2396  psi-mi:hs90a_human(display_short)|psi-mi:Heat ...   
2397  psi-mi:nsf_human(display_short)|psi-mi:Vesicle...   
2398  psi-mi:mark3_human(display_short)|psi-mi:MAP/m...   
2399  psi-mi:dyn1_human(display_short)|psi-mi:Dynami...   

                        Interaction detection method(s)  \
0      psi-mi:"MI:0007"(anti tag coimmunoprecipitation)   
1      psi-mi:"MI:0007"(anti tag coimmunoprecipitation)   
2      psi-mi:"MI:0007"(anti tag coimmunoprecipitation)   
3      psi-mi:"MI:0007"(anti tag coimmunoprecipitation)   
4      psi-mi:"MI:0007"(anti tag coimmunoprecipitation)   
...                                                 ...   
2395  psi-mi:"MI:0006"(anti bait coimmunoprecipitation)   
2396  psi-mi:"MI:0006"(anti bait coimmunoprecipitation)   
2397  psi-mi:"MI:0006"(anti bait coimmunoprecipitation)   
23

In [176]:
gene_dataset['Interactor A P_id'] = list(map(lambda x: (x.split(':')[1]).split('-')[0], gene_dataset['# ID(s) interactor A'].tolist()))

In [33]:
first_second_level_matched_genes =  p_g_new[p_g_new['converted_alias'].isin(second_level_matched_genes)]

In [148]:
first_second_level_matched_genes

initial_alias  converted_alias    name  \
0           P05067  ENSG00000142192     APP   
30          P49768  ENSG00000080815   PSEN1   
104         P54253  ENSG00000124788   ATXN1   
151         P42858  ENSG00000197386     HTT   
270         O14672  ENSG00000137845  ADAM10   
...            ...              ...     ...   
3220        Q02410  ENSG00000107282   APBA1   
3225        P49407  ENSG00000137486   ARRB1   
3235        P32121  ENSG00000141480   ARRB2   
3245        P53805  ENSG00000159200   RCAN1   
3264        Q96L34  ENSG00000007047   MARK4   

                                            description  \
0     amyloid beta precursor protein [Source:HGNC Sy...   
30      presenilin 1 [Source:HGNC Symbol;Acc:HGNC:9508]   
104        ataxin 1 [Source:HGNC Symbol;Acc:HGNC:10548]   
151       huntingtin [Source:HGNC Symbol;Acc:HGNC:4851]   
270   ADAM metallopeptidase domain 10 [Source:HGNC S...   
...                                                 ...   
3220  amyloid beta precursor protein binding family ...   
3225  arrestin beta 1 [Source:HGNC Symbol;Acc:HGNC:711]   
3235  arrestin beta 2 [Source:HGNC Symbol;Acc:HGNC:712]   
3245  regulator of calcineurin 1 [Source:HGNC Symbol...   
3264  microtubule affinity regulating kinase 4 [Sour...   

                                namespace  
0     UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
30    UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
104   UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
151   UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
270   UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
...                                   ...  
3220  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
3225  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
3235  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
3245  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
3264  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  

[133 rows x 5 columns]

In [149]:
first_second_level_matched_genes_list = first_second_level_matched_genes['initial_alias'].unique().tolist()

In [189]:
first_second_level_matched_ens_list = first_second_level_matched_genes['converted_alias'].unique().tolist()

In [171]:
len(first_second_level_matched_genes['converted_alias'].unique().tolist())

133

In [153]:
len(first_second_level_matched_genes_list)

131

##### 2 genes of first_secong level matched genes (133) are missing in gene dataset

In [180]:
first_second_level_matched_genes.head()

initial_alias  converted_alias    name  \
0          P05067  ENSG00000142192     APP   
30         P49768  ENSG00000080815   PSEN1   
104        P54253  ENSG00000124788   ATXN1   
151        P42858  ENSG00000197386     HTT   
270        O14672  ENSG00000137845  ADAM10   

                                           description  \
0    amyloid beta precursor protein [Source:HGNC Sy...   
30     presenilin 1 [Source:HGNC Symbol;Acc:HGNC:9508]   
104       ataxin 1 [Source:HGNC Symbol;Acc:HGNC:10548]   
151      huntingtin [Source:HGNC Symbol;Acc:HGNC:4851]   
270  ADAM metallopeptidase domain 10 [Source:HGNC S...   

                               namespace  
0    UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
30   UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
104  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
151  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC  
270  UNIPROTSWISSPROT_ACC,UNIPROT_GN_ACC

In [183]:
gene_dataset['initial_alias'] = gene_dataset['Interactor A P_id']

In [187]:
gene_dataset_with_ens = gene_dataset.join(first_second_level_matched_genes[['initial_alias','converted_alias']].set_index('initial_alias'),on='initial_alias')

In [190]:
gene_dataset_with_ens.head()

# ID(s) interactor A ID(s) interactor B  \
0     uniprotkb:P05067   uniprotkb:O95864   
1     uniprotkb:P05067   uniprotkb:P13861   
2     uniprotkb:P05067   uniprotkb:Q8TD31   
3     uniprotkb:P05067   uniprotkb:Q15392   
4     uniprotkb:P05067   uniprotkb:Q15125   

                             Alt. ID(s) interactor A  \
0  ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
1  ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
2  ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
3  ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   
4  ensembl:ENSP00000284981.4|uniprotkb:P09000|uni...   

                             Alt. ID(s) interactor B  \
0  uniprotkb:A8K2M6|uniprotkb:Q6MZQ7|uniprotkb:Q9...   
1  uniprotkb:Q16823|intact:EBI-6116682|uniprotkb:...   
2  uniprotkb:Q9NRK8|uniprotkb:Q9NWY9|uniprotkb:Q9...   
3  uniprotkb:D3DQ51|uniprotkb:Q9HBA8|uniprotkb:B7...   
4  uniprotkb:Q6IBI9|uniprotkb:Q6FGL3|ensembl:ENSP...   

                              Alias(es) interactor A  \
0  psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
1  psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
2  psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
3  psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   
4  psi-mi:a4_human(display_short)|psi-mi:Amyloid-...   

                              Alias(es) interactor B  \
0  psi-mi:fads2_human(display_short)|psi-mi:Acyl-...   
1  psi-mi:kap2_human(display_short)|psi-mi:cAMP-d...   
2  psi-mi:cchcr_human(display_short)|psi-mi:Coile...   
3  psi-mi:dhc24_human(display_short)|psi-mi:"Delt...   
4  psi-mi:ebp_human(display_short)|psi-mi:"3-beta...   

                    Interaction detection method(s) Publication 1st author(s)  \
0  psi-mi:"MI:0007"(anti tag coimmunoprecipitation)      Hosp F. et al.(2015)   
1  psi-mi:"MI:0007"(anti tag coimmunoprecipitation)      Hosp F. et al.(2015)   
2  psi-mi:"MI:0007"(anti tag coimmunoprecipitation)      Hosp F. et al.(2015)   
3  psi-mi:"MI:0007"(anti tag coimmunoprecipitation)      Hosp F. et al.(2015)   
4  psi-mi:"MI:0007"(anti tag coimmunoprecipitation)      Hosp F. et al.(2015)   

                           Publication Identifier(s)  \
0  intact:EBI-21132292|pubmed:25959826|imex:IM-27137   
1  intact:EBI-21132292|pubmed:25959826|imex:IM-27137   
2  intact:EBI-21132292|pubmed:25959826|imex:IM-27137   
3  intact:EBI-21132292|pubmed:25959826|imex:IM-27137   
4  intact:EBI-21132292|pubmed:25959826|imex:IM-27137   

                           Taxid interactor A  \
0  taxid:9606(human)|taxid:9606(Homo sapiens)   
1  taxid:9606(human)|taxid:9606(Homo sapiens)   
2  taxid:9606(human)|taxid:9606(Homo sapiens)   
3  taxid:9606(human)|taxid:9606(Homo sapiens)   
4  taxid:9606(human)|taxid:9606(Homo sapiens)   

                           Taxid interactor B            Interaction type(s)  \
0  taxid:9606(human)|taxid:9606(Homo sapiens)  psi-mi:"MI:0914"(association)   
1  taxid:9606(human)|taxid:9606(Homo sapiens)  psi-mi:"MI:0914"(association)   
2  taxid:9606(human)|taxid:9606(Homo sapiens)  psi-mi:"MI:0914"(association)   
3  taxid:9606(human)|taxid:9606(Homo sapiens)  psi-mi:"MI:0914"(association)   
4  taxid:9606(human)|taxid:9606(Homo sapiens)  psi-mi:"MI:0914"(association)   

          Source database(s)            Interaction identifier(s)  \
0  psi-mi:"MI:0486"(UniProt)  intact:EBI-21132574|imex:IM-27137-1   
1  psi-mi:"MI:0486"(UniProt)  intact:EBI-21132574|imex:IM-27137-1   
2  psi-mi:"MI:0486"(UniProt)  intact:EBI-21132574|imex:IM-27137-1   
3  psi-mi:"MI:0486"(UniProt)  intact:EBI-21132574|imex:IM-27137-1   
4  psi-mi:"MI:0486"(UniProt)  intact:EBI-21132574|imex:IM-27137-1   

  Confidence value(s) Interactor A P_id initial_alias  converted_alias  
0                   -            P05067        P05067  ENSG00000142192  
1                   -            P05067        P05067  ENSG00000142192  
2                   -            P05067        P05067  ENSG00000142192  
3                   -            P05067        P05067  ENSG00000142192  
4  

In [192]:
second_third_level_dataset = gene_dataset_with_ens[gene_dataset_with_ens['converted_alias'].isin(first_second_level_matched_ens_list)]

In [195]:
len(second_third_level_dataset['converted_alias'].unique().tolist())

133

In [196]:
len(second_third_level_dataset['ID(s) interactor B'].unique().tolist())

1274

In [197]:
adj_matrix_second_third = pd.crosstab(second_third_level_dataset['converted_alias'],second_third_level_dataset['ID(s) interactor B'])

In [206]:
adj_matrix_second_third.head()

ID(s) interactor B  intact:EBI-11820565  intact:EBI-20723938  \
converted_alias                                                
ENSG00000002822                       0                    0   
ENSG00000003393                       0                    0   
ENSG00000005884                       0                    0   
ENSG00000007047                       0                    0   
ENSG00000010278                       0                    0   

ID(s) interactor B  intact:EBI-26519605  uniprotkb:A0A024R0E2  \
converted_alias                                                 
ENSG00000002822                       0                     0   
ENSG00000003393                       0                     0   
ENSG00000005884                       0                     0   
ENSG00000007047                       0                     0   
ENSG00000010278                       0                     0   

ID(s) interactor B  uniprotkb:A0A024R0H7  uniprotkb:A0A024R0M6  \
converted_alias                                                  
ENSG00000002822                        0                     0   
ENSG00000003393                        0                     0   
ENSG00000005884                        0                     0   
ENSG00000007047                        0                     0   
ENSG00000010278                        0                     0   

ID(s) interactor B  uniprotkb:A0A024R1K7  uniprotkb:A0A024R1U0  \
converted_alias                                                  
ENSG00000002822                        0                     0   
ENSG00000003393                        0                     0   
ENSG00000005884                        0                     0   
ENSG00000007047                        0                     0   
ENSG00000010278                        0                     0   

ID(s) interactor B  uniprotkb:A0A024R1U4  uniprotkb:A0A024R1Z6  ...  \
converted_alias                                                 ...   
ENSG00000002822                        0                     0  ...   
ENSG00000003393                        0                     0  ...   
ENSG00000005884                        0                     0  ...   
ENSG00000007047                        0                     0  ...   
ENSG00000010278                        0                     0  ...   

ID(s) interactor B  uniprotkb:Q9Y6Y0  uniprotkb:R9S3C3  uniprotkb:V5YQL4  \
converted_alias                                                            
ENSG00000002822                    0                 0                 0   
ENSG00000003393                    0                 0                 0   
ENSG00000005884                    0                 0                 0   
ENSG00000007047                    0                 0                 0   
ENSG00000010278                    0                 0                 0   

ID(s) interactor B  uniprotkb:V9HW22  uniprotkb:V9HW31  uniprotkb:V9HW96  \
converted_alias                                                            
ENSG00000002822                    0                 0                 0   
ENSG00000003393                    0                 0                 0   
ENSG00000005884                    0                 0                 0   
ENSG00000007047                    0                 0                 0   
ENSG00000010278                    0                 0                 0   

ID(s) interactor B  uniprotkb:V9HWB4  uniprotkb:V9HWB8  uniprotkb:V9HWC2  \
converted_alias                                                            
ENSG00000002822                    0                 0                 0   
ENSG00000003393                    0                 0                 0   
ENSG00000005884                    0                 0                 0   
ENSG00000007047                    0                 0                 0   
ENSG00000010278                    0                 0                 0   

ID(s) interactor B  uniprotkb:V9HWH9  
converted_alias                  

In [213]:
len(set(adj_matrix_first_second.columns.tolist()).intersection(adj_matrix_second_third.index.tolist()))

133

In [215]:
adj_matrix_first_second.head()

gene_id       ENSG00000002822  ENSG00000003393  ENSG00000005884  \
SNP_id                                                            
10:102764091                0                0                0   
10:105221736                0                0                0   
10:112170486                0                0                0   
10:112558236                0                0                0   
10:112650209                0                0                0   

gene_id       ENSG00000007047  ENSG00000010278  ENSG00000010810  \
SNP_id                                                            
10:102764091                0                0                0   
10:105221736                0                0                0   
10:112170486                0                0                0   
10:112558236                0                0                0   
10:112650209                0                0                0   

gene_id       ENSG00000023191  ENSG00000034693  ENSG00000047457  \
SNP_id                                                            
10:102764091                0                0                0   
10:105221736                0                0                0   
10:112170486                0                0                0   
10:112558236                0                0                0   
10:112650209                0                0                0   

gene_id       ENSG00000051128  ...  ENSG00000197122  ENSG00000197386  \
SNP_id                         ...                                     
10:102764091                0  ...                0                0   
10:105221736                0  ...                0                0   
10:112170486                0  ...                0                0   
10:112558236                0  ...                0                0   
10:112650209                0  ...                0                0   

gene_id       ENSG00000198087  ENSG00000198736  ENSG00000205155  \
SNP_id                                                            
10:102764091                0                0                0   
10:105221736                0                0                0   
10:112170486                0                0                0   
10:112558236                0                0                0   
10:112650209                0                0                0   

gene_id       ENSG00000205571  ENSG00000214253  ENSG00000221823  \
SNP_id                                                            
10:102764091                0                0                0   
10:105221736                0                0                0   
10:112170486                0                0                0   
10:112558236                0                0                0   
10:112650209                0                0                0   

gene_id       ENSG00000221829  ENSG00000227544  
SNP_id                                          
10:102764091                0                0  
10:105221736                0                0  
10:112170486                0                0  
10:112558236                0                0  
10:112650209                0                0  

[5 rows x 133 columns]

In [216]:
adj_matrix_first_second.to_csv('adj_matrix_first_second.csv')

In [222]:
pd.read_csv('adj_matrix_first_second.csv',index_col = 'SNP_id').T

SNP_id           10:102764091  10:105221736  10:112170486  10:112558236  \
ENSG00000002822             0             0             0             0   
ENSG00000003393             0             0             0             0   
ENSG00000005884             0             0             0             0   
ENSG00000007047             0             0             0             0   
ENSG00000010278             0             0             0             0   
...                       ...           ...           ...           ...   
ENSG00000205571             0             0             0             0   
ENSG00000214253             0             0             0             0   
ENSG00000221823             0             0             0             0   
ENSG00000221829             0             0             0             0   
ENSG00000227544             0             0             0             0   

SNP_id           10:112650209  10:112802423  10:113118382  10:113125652  \
ENSG00000002822             0             0             0             0   
ENSG00000003393             0             0             0             0   
ENSG00000005884             0             0             0             0   
ENSG00000007047             0             0             0             0   
ENSG00000010278             0             0             0             0   
...                       ...           ...           ...           ...   
ENSG00000205571             0             0             0             0   
ENSG00000214253             0             0             0             0   
ENSG00000221823             0             0             0             0   
ENSG00000221829             0             0             0             0   
ENSG00000227544             0             0             0             0   

SNP_id           10:113489179  10:118984464  ...  9:27558437  9:27573897  \
ENSG00000002822             0             0  ...           0           0   
ENSG00000003393             0             0  ...           0           0   
ENSG00000005884             0             0  ...           0           0   
ENSG00000007047             0             0  ...           0           0   
ENSG00000010278             0             0  ...           0           0   
...                       ...           ...  ...         ...         ...   
ENSG00000205571             0             0  ...           0           0   
ENSG00000214253             0             0  ...           0           0   
ENSG00000221823             0             0  ...           0           0   
ENSG00000221829             0             0  ...           0           0   
ENSG00000227544             0             0  ...           0           0   

SNP_id           9:27578935  9:27583817  9:27586451  9:27594491  9:28387593  \
ENSG00000002822           0           0           0           0           0   
ENSG00000003393           0           0           0           0           0   
ENSG00000005884           0           0           0           0           0   
ENSG00000007047           0           0           0           0           0   
ENSG00000010278           0           0           0           0           0   
...                     ...         ...         ...         ...         ...   
ENSG00000205571           0           0           0           0           0   
ENSG00000214253           0           0           0           0           0   
ENSG00000221823           0           0           0           0           0   
ENSG00000221829           0           0           0           0           0   
ENSG00000227544           0           0           0           0           0   

SNP_id           9:35022891  9:72014104  9:72128165  
ENSG00000002822           0           0           0  
ENSG00000003393           0           0           0  
ENSG00000005884           0           0           0  
ENSG00000007047           0           0           0  
ENSG00000010278           0           0           0  
...         

In [219]:
adj_matrix_second_third = adj_matrix_second_third.loc[adj_matrix_first_second.columns.tolist()]

In [220]:
adj_matrix_second_third.to_csv('adj_matrix_second_third.csv')

In [225]:
pd.read_csv('adj_matrix_second_third.csv',index_col = 'converted_alias').T

converted_alias       ENSG00000002822  ENSG00000003393  ENSG00000005884  \
intact:EBI-11820565                 0                0                0   
intact:EBI-20723938                 0                0                0   
intact:EBI-26519605                 0                0                0   
uniprotkb:A0A024R0E2                0                0                0   
uniprotkb:A0A024R0H7                0                0                0   
...                               ...              ...              ...   
uniprotkb:V9HW96                    0                0                0   
uniprotkb:V9HWB4                    0                0                0   
uniprotkb:V9HWB8                    0                0                0   
uniprotkb:V9HWC2                    0                0                0   
uniprotkb:V9HWH9                    0                0                0   

converted_alias       ENSG00000007047  ENSG00000010278  ENSG00000010810  \
intact:EBI-11820565                 0                0                0   
intact:EBI-20723938                 0                0                0   
intact:EBI-26519605                 0                0                0   
uniprotkb:A0A024R0E2                0                0                0   
uniprotkb:A0A024R0H7                0                0                0   
...                               ...              ...              ...   
uniprotkb:V9HW96                    0                0                0   
uniprotkb:V9HWB4                    0                0                0   
uniprotkb:V9HWB8                    0                0                0   
uniprotkb:V9HWC2                    0                0                0   
uniprotkb:V9HWH9                    0                0                0   

converted_alias       ENSG00000023191  ENSG00000034693  ENSG00000047457  \
intact:EBI-11820565                 0                0                0   
intact:EBI-20723938                 0                0                0   
intact:EBI-26519605                 0                0                0   
uniprotkb:A0A024R0E2                0                0                0   
uniprotkb:A0A024R0H7                0                0                0   
...                               ...              ...              ...   
uniprotkb:V9HW96                    0                0                0   
uniprotkb:V9HWB4                    0                0                0   
uniprotkb:V9HWB8                    0                0                0   
uniprotkb:V9HWC2                    0                0                0   
uniprotkb:V9HWH9                    0                0                0   

converted_alias       ENSG00000051128  ...  ENSG00000197122  ENSG00000197386  \
intact:EBI-11820565                 0  ...                0                1   
intact:EBI-20723938                 0  ...                0                0   
intact:EBI-26519605                 0  ...                0                0   
uniprotkb:A0A024R0E2                0  ...                0                0   
uniprotkb:A0A024R0H7                0  ...                0                0   
...                               ...  ...              ...              ...   
uniprotkb:V9HW96                    0  ...                0                0   
uniprotkb:V9HWB4                    0  ...                0                0   
uniprotkb:V9HWB8                    0  ...                0                0   
uniprotkb:V9HWC2                    0  ...                0                0   
uniprotkb:V9HWH9                    0  ...                0                0   

converted_alias       ENSG00000198087  ENSG00000198736  ENSG00000205155  \
intact:EBI-11820565                 0                0                0   
intact:EBI-20723938                 0                0                0   
intact:EBI-26519605                 0                0                0   
uniprotkb:A0A024R0E2                

## Creating the dataset for Pytorch

In [65]:
df_list = [pd.read_csv(file,sep=' ',index_col='FID',usecols=lambda x: x not in ['PHENOTYPE','SEX','IID','PAT','MAT']) for file in csv_flies]

In [66]:
big_df = pd.concat(df_list,axis=1)

In [67]:
big_df.shape

(36620, 37606)

In [68]:
big_df.index

Int64Index([1767364, 2268235, 4521147, 5278224, 1925234, 5738107, 2218042,
            2405794, 4423490, 5773801,
            ...
            1202774, 1126932, 2754094, 3545763, 6000176, 4283752, 4622763,
            2870611, 5854257, 1107612],
           dtype='int64', name='FID', length=36620)

In [74]:
rs_ids = list(map(lambda x: x.split('_')[0],big_df.columns.tolist()))

In [75]:
big_df.columns = rs_ids

In [144]:
first_matched_rsids = first_second_level_genes.rs_id.unique().tolist()

In [145]:
len(first_matched_rsids)

400

In [146]:
selected_snps_dataset = big_df[first_matched_rsids]

In [147]:
selected_snps_dataset = selected_snps_dataset.loc[:,~selected_snps_dataset.columns.duplicated()]

#### Matching subjects with working memory data

In [91]:
label_file= '/data/users2/pnadigapusuresh1/Projects/ukbiobank/Data/subset_vars.csv'

In [92]:
vars = pd.read_csv(label_file,index_col='eid',
                            usecols=['eid','maximum_digits_remembered_correctly_f4282_2_0',
                            'sex_f31_0_0','age_when_attended_assessment_centre_f21003_0_0'])

In [93]:
vars.columns = ['sex','score','age']

In [94]:
vars

sex  score   age
eid                      
5230587  1.0    8.0  57.0
2369031  0.0    7.0  45.0
4304744  0.0    7.0  69.0
4373404  1.0    6.0  51.0
1229224  1.0    8.0  60.0
...      ...    ...   ...
3001470  0.0    4.0  46.0
2677873  1.0    NaN  62.0
5343321  0.0    6.0  56.0
2504488  1.0    9.0  50.0
4348663  1.0    7.0  64.0

[39760 rows x 3 columns]

In [95]:
vars = vars.loc[vars.score.isin([2,3,4,5,9,10,11,12])]

In [105]:
selected_snps_dataset

rs76332820  rs4944873  rs7131683  rs139848901  rs10793118  \
FID                                                                  
1767364           0        0.0        0.0          1.0         1.0   
2268235           0        1.0        0.0          0.0         2.0   
4521147           0        1.0        0.0          0.0         2.0   
5278224           0        1.0        0.0          0.0         1.0   
1925234           0        2.0        1.0          0.0         1.0   
...             ...        ...        ...          ...         ...   
4283752           1        0.0        0.0          0.0         2.0   
4622763           0        1.0        1.0          0.0         1.0   
2870611           0        0.0        0.0          0.0         0.0   
5854257           0        1.0        0.0          0.0         1.0   
1107612           1        NaN        0.0          0.0         0.0   

         rs145191148  rs672701  rs12113715  rs186098536  rs6962921  ...  \
FID                                                                 ...   
1767364          0.0       0.0         0.0          0.0        1.0  ...   
2268235          0.0       2.0         0.0          0.0        0.0  ...   
4521147          0.0       0.0         2.0          0.0        1.0  ...   
5278224          0.0       0.0         2.0          0.0        0.0  ...   
1925234          0.0       0.0         1.0          0.0        1.0  ...   
...              ...       ...         ...          ...        ...  ...   
4283752          0.0       1.0         1.0          0.0        1.0  ...   
4622763          0.0       0.0         1.0          0.0        1.0  ...   
2870611          0.0       0.0         1.0          0.0        1.0  ...   
5854257          1.0       0.0         0.0          0.0        2.0  ...   
1107612          0.0       0.0         1.0          0.0        1.0  ...   

         rs12981911  rs8102125  rs11881662  rs75332038  rs310434  rs77980659  \
FID                                                                            
1767364         1.0        1.0         0.0         0.0       1.0         0.0   
2268235         1.0        0.0         1.0         0.0       0.0         0.0   
4521147         1.0        0.0         1.0         0.0       0.0         0.0   
5278224         2.0        1.0         0.0         0.0       0.0         0.0   
1925234         0.0        1.0         1.0         0.0       1.0         0.0   
...             ...        ...         ...         ...       ...         ...   
4283752         1.0        1.0         0.0         0.0       1.0         0.0   
4622763         0.0        1.0         0.0         1.0       1.0         0.0   
2870611         1.0        0.0         0.0         NaN       NaN         0.0   
5854257         1.0        1.0         0.0         0.0       1.0         0.0   
1107612         1.0        1.0         1.0         0.0       0.0         0.0   

         rs4447797  rs13062950  rs9839511  rs145914525  
FID                                                     
1767364        0.0         0.0        0.0          0.0  
2268235        2.0         1.0        0.0          0.0  
4521147        2.0         0.0        0.0          0.0  
5278224        1.0         2.0        1.0          0.0  
1925234        2.0         2.0        0.0          0.0  
...            ...         ...        ...          ...  
4283752        2.0         2.0        0.0          0.0  
4622763        1.0         2.0        1.0          0.0  
2870611        1.0         1.0        1.0          0.0  
5854257        0.0         0.0        0.0          0.0  
1107612        1.0         1.0        1.0          0.0  

[36620 rows x 400 columns]

In [117]:
subjects_with_working_mem_and_gene = set(vars.index.tolist()).intersection(selected_snps_dataset.index.tolist())

In [119]:
gene_working_mem = selected_snps_dataset.loc[list(subjects_with_working_mem_and_gene)]

In [120]:
gene_working_mem

rs76332820  rs4944873  rs7131683  rs139848901  rs10793118  \
FID                                                                  
2588677           0        0.0        1.0          0.0         2.0   
3670029           0        0.0        0.0          0.0         1.0   
1245201           1        1.0        0.0          0.0         1.0   
2162713           0        2.0        0.0          0.0         0.0   
5636125           0        2.0        0.0          0.0         1.0   
...             ...        ...        ...          ...         ...   
4259788           0        2.0        0.0          0.0         0.0   
5275598           0        1.0        0.0          1.0         1.0   
3243989           0        0.0        0.0          0.0         NaN   
1343446           0        1.0        1.0          0.0         1.0   
1671167           0        1.0        0.0          0.0         1.0   

         rs145191148  rs672701  rs12113715  rs186098536  rs6962921  ...  \
FID                                                                 ...   
2588677          0.0       2.0         1.0          0.0        2.0  ...   
3670029          0.0       1.0         1.0          0.0        2.0  ...   
1245201          0.0       1.0         1.0          0.0        2.0  ...   
2162713          0.0       0.0         2.0          0.0        2.0  ...   
5636125          0.0       0.0         1.0          0.0        0.0  ...   
...              ...       ...         ...          ...        ...  ...   
4259788          0.0       0.0         1.0          0.0        1.0  ...   
5275598          0.0       1.0         1.0          0.0        1.0  ...   
3243989          0.0       0.0         2.0          0.0        1.0  ...   
1343446          0.0       0.0         1.0          0.0        1.0  ...   
1671167          0.0       0.0         NaN          0.0        0.0  ...   

         rs12981911  rs8102125  rs11881662  rs75332038  rs310434  rs77980659  \
FID                                                                            
2588677         2.0        1.0         0.0         0.0       1.0         1.0   
3670029         1.0        0.0         1.0         0.0       0.0         0.0   
1245201         1.0        0.0         0.0         0.0       1.0         0.0   
2162713         1.0        1.0         1.0         0.0       0.0         0.0   
5636125         1.0        1.0         0.0         0.0       0.0         0.0   
...             ...        ...         ...         ...       ...         ...   
4259788         1.0        0.0         1.0         0.0       1.0         0.0   
5275598         0.0        0.0         0.0         0.0       0.0         0.0   
3243989         0.0        1.0         0.0         0.0       1.0         0.0   
1343446         1.0        1.0         0.0         0.0       1.0         0.0   
1671167         0.0        2.0         0.0         0.0       0.0         0.0   

         rs4447797  rs13062950  rs9839511  rs145914525  
FID                                                     
2588677        0.0         2.0        0.0          0.0  
3670029        1.0         2.0        0.0          0.0  
1245201        1.0         1.0        0.0          0.0  
2162713        0.0         0.0        0.0          0.0  
5636125        0.0         2.0        0.0          1.0  
...            ...         ...        ...          ...  
4259788        1.0         2.0        0.0          0.0  
5275598        0.0         0.0        0.0          0.0  
3243989        1.0         1.0        0.0          0.0  
1343446        1.0         2.0        0.0          1.0  
1671167        1.0         1.0        0.0          0.0  

[5002 rows x 400 columns]

In [121]:
gene_working_mem.to_csv('gene_working_mem_dataset.csv')

In [136]:
gene_working_mem = gene_working_mem.fillna(0)

In [137]:
gene_working_mem

rs76332820  rs4944873  rs7131683  rs139848901  rs10793118  \
FID                                                                  
2588677           0        0.0        1.0          0.0         2.0   
3670029           0        0.0        0.0          0.0         1.0   
1245201           1        1.0        0.0          0.0         1.0   
2162713           0        2.0        0.0          0.0         0.0   
5636125           0        2.0        0.0          0.0         1.0   
...             ...        ...        ...          ...         ...   
4259788           0        2.0        0.0          0.0         0.0   
5275598           0        1.0        0.0          1.0         1.0   
3243989           0        0.0        0.0          0.0         0.0   
1343446           0        1.0        1.0          0.0         1.0   
1671167           0        1.0        0.0          0.0         1.0   

         rs145191148  rs672701  rs12113715  rs186098536  rs6962921  ...  \
FID                                                                 ...   
2588677          0.0       2.0         1.0          0.0        2.0  ...   
3670029          0.0       1.0         1.0          0.0        2.0  ...   
1245201          0.0       1.0         1.0          0.0        2.0  ...   
2162713          0.0       0.0         2.0          0.0        2.0  ...   
5636125          0.0       0.0         1.0          0.0        0.0  ...   
...              ...       ...         ...          ...        ...  ...   
4259788          0.0       0.0         1.0          0.0        1.0  ...   
5275598          0.0       1.0         1.0          0.0        1.0  ...   
3243989          0.0       0.0         2.0          0.0        1.0  ...   
1343446          0.0       0.0         1.0          0.0        1.0  ...   
1671167          0.0       0.0         0.0          0.0        0.0  ...   

         rs12981911  rs8102125  rs11881662  rs75332038  rs310434  rs77980659  \
FID                                                                            
2588677         2.0        1.0         0.0         0.0       1.0         1.0   
3670029         1.0        0.0         1.0         0.0       0.0         0.0   
1245201         1.0        0.0         0.0         0.0       1.0         0.0   
2162713         1.0        1.0         1.0         0.0       0.0         0.0   
5636125         1.0        1.0         0.0         0.0       0.0         0.0   
...             ...        ...         ...         ...       ...         ...   
4259788         1.0        0.0         1.0         0.0       1.0         0.0   
5275598         0.0        0.0         0.0         0.0       0.0         0.0   
3243989         0.0        1.0         0.0         0.0       1.0         0.0   
1343446         1.0        1.0         0.0         0.0       1.0         0.0   
1671167         0.0        2.0         0.0         0.0       0.0         0.0   

         rs4447797  rs13062950  rs9839511  rs145914525  
FID                                                     
2588677        0.0         2.0        0.0          0.0  
3670029        1.0         2.0        0.0          0.0  
1245201        1.0         1.0        0.0          0.0  
2162713        0.0         0.0        0.0          0.0  
5636125        0.0         2.0        0.0          1.0  
...            ...         ...        ...          ...  
4259788        1.0         2.0        0.0          0.0  
5275598        0.0         0.0        0.0          0.0  
3243989        1.0         1.0        0.0          0.0  
1343446        1.0         2.0        0.0          1.0  
1671167        1.0         1.0        0.0          0.0  

[5002 rows x 400 columns]

In [124]:
pd.read_csv('gene_working_mem_dataset.csv',index_col='FID')

rs76332820  rs4944873  rs7131683  rs139848901  rs10793118  \
FID                                                                  
2588677           0        0.0        1.0          0.0         2.0   
3670029           0        0.0        0.0          0.0         1.0   
1245201           1        1.0        0.0          0.0         1.0   
2162713           0        2.0        0.0          0.0         0.0   
5636125           0        2.0        0.0          0.0         1.0   
...             ...        ...        ...          ...         ...   
4259788           0        2.0        0.0          0.0         0.0   
5275598           0        1.0        0.0          1.0         1.0   
3243989           0        0.0        0.0          0.0         NaN   
1343446           0        1.0        1.0          0.0         1.0   
1671167           0        1.0        0.0          0.0         1.0   

         rs145191148  rs672701  rs12113715  rs186098536  rs6962921  ...  \
FID                                                                 ...   
2588677          0.0       2.0         1.0          0.0        2.0  ...   
3670029          0.0       1.0         1.0          0.0        2.0  ...   
1245201          0.0       1.0         1.0          0.0        2.0  ...   
2162713          0.0       0.0         2.0          0.0        2.0  ...   
5636125          0.0       0.0         1.0          0.0        0.0  ...   
...              ...       ...         ...          ...        ...  ...   
4259788          0.0       0.0         1.0          0.0        1.0  ...   
5275598          0.0       1.0         1.0          0.0        1.0  ...   
3243989          0.0       0.0         2.0          0.0        1.0  ...   
1343446          0.0       0.0         1.0          0.0        1.0  ...   
1671167          0.0       0.0         NaN          0.0        0.0  ...   

         rs12981911  rs8102125  rs11881662  rs75332038  rs310434  rs77980659  \
FID                                                                            
2588677         2.0        1.0         0.0         0.0       1.0         1.0   
3670029         1.0        0.0         1.0         0.0       0.0         0.0   
1245201         1.0        0.0         0.0         0.0       1.0         0.0   
2162713         1.0        1.0         1.0         0.0       0.0         0.0   
5636125         1.0        1.0         0.0         0.0       0.0         0.0   
...             ...        ...         ...         ...       ...         ...   
4259788         1.0        0.0         1.0         0.0       1.0         0.0   
5275598         0.0        0.0         0.0         0.0       0.0         0.0   
3243989         0.0        1.0         0.0         0.0       1.0         0.0   
1343446         1.0        1.0         0.0         0.0       1.0         0.0   
1671167         0.0        2.0         0.0         0.0       0.0         0.0   

         rs4447797  rs13062950  rs9839511  rs145914525  
FID                                                     
2588677        0.0         2.0        0.0          0.0  
3670029        1.0         2.0        0.0          0.0  
1245201        1.0         1.0        0.0          0.0  
2162713        0.0         0.0        0.0          0.0  
5636125        0.0         2.0        0.0          1.0  
...            ...         ...        ...          ...  
4259788        1.0         2.0        0.0          0.0  
5275598        0.0         0.0        0.0          0.0  
3243989        1.0         1.0        0.0          0.0  
1343446        1.0         2.0        0.0          1.0  
1671167        1.0         1.0        0.0          0.0  

[5002 rows x 400 columns]

In [132]:
vars = vars.loc[gene_working_mem.index.tolist()]

In [133]:
vars.iloc[200]

sex       0.0
score     4.0
age      53.0
Name: 4228319, dtype: float64

In [138]:
gene_working_mem.iloc[200].to_numpy()

array([0., 1., 0., 0., 0., 0., 0., 1., 0., 1., 0., 0., 0., 2., 1., 1., 0.,
       0., 0., 0., 1., 0., 0., 1., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0.,
       0., 0., 0., 2., 2., 1., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0.,
       0., 1., 1., 0., 0., 0., 0., 1., 1., 1., 1., 0., 1., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 1., 1., 1., 1.,
       0., 1., 0., 1., 1., 0., 1., 0., 0., 0., 2., 1., 0., 1., 0., 0., 1.,
       1., 0., 1., 1., 0., 0., 0., 0., 0., 1., 0., 0., 0., 2., 0., 0., 1.,
       1., 0., 0., 0., 1., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1.,
       0., 0., 1., 0., 0., 0., 1., 0., 0., 1., 0., 1., 0., 0., 0., 0., 2.,
       0., 0., 2., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 2., 1., 0., 0.,
       1., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 2., 0., 0., 1., 0., 1.,
       0., 1., 0., 0., 1., 0., 2., 0., 0., 0., 1., 1., 0., 0., 0., 2., 0.,
       0., 0., 1., 0., 2., 0., 2., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.,
       0., 0., 0., 0., 0.

In [139]:
import torch

In [140]:
torch.tensor(gene_working_mem.iloc[200].to_numpy())

tensor([0., 1., 0., 0., 0., 0., 0., 1., 0., 1., 0., 0., 0., 2., 1., 1., 0., 0.,
        0., 0., 1., 0., 0., 1., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
        0., 2., 2., 1., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 1., 1.,
        0., 0., 0., 0., 1., 1., 1., 1., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0.,
        0., 1., 0., 0., 0., 1., 0., 0., 0., 1., 1., 1., 1., 0., 1., 0., 1., 1.,
        0., 1., 0., 0., 0., 2., 1., 0., 1., 0., 0., 1., 1., 0., 1., 1., 0., 0.,
        0., 0., 0., 1., 0., 0., 0., 2., 0., 0., 1., 1., 0., 0., 0., 1., 0., 1.,
        0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 0., 0., 1., 0., 0., 0., 1., 0.,
        0., 1., 0., 1., 0., 0., 0., 0., 2., 0., 0., 2., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 2., 1., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0., 0., 0.,
        1., 2., 0., 0., 1., 0., 1., 0., 1., 0., 0., 1., 0., 2., 0., 0., 0., 1.,
        1., 0., 0., 0., 2., 0., 0., 0., 1., 0., 2., 0., 2., 0., 0., 0., 0., 0.,
        0., 1., 0., 0., 0., 0., 0., 0., 